### ISSUE TITLE
[Feature Request] Semantic highlighting for the various uses of square brackets

### ISSUE THREAD

**Issue by nmsmith** [modularml/mojo#224]
## Request

Square brackets have three different uses in Mojo:
- Array literals
- Indexing (which can be made available on user-defined data types via `__getitem__`)
- Parameters

I request that Mojo's language server, and its VSCode extension, use semantic highlighting to clearly distinguish each of these meanings.

### Motivation

When skimming a Mojo codebase, one is going to see a lot of `[..]`. Each occurrence will have one of three possible meanings. It is crucial to be able to quickly figuring out _which_ of these meanings is the correct one. This could end up being particularly difficult in a region of code that invokes several parametric functions and also does some work with arrays or dictionaries.

Also, the different uses of `[..]` are likely to cause a lot of confusion for people who are attempting to learn Mojo as a first language. Colorization would help learners realize that they are looking at something different. For example, if a learner has just finished learning about arrays, and array literals always have **yellow brackets**, they will come to associate arrays with the color yellow. Then, if they later see a type such as `Tree[Int]`, but **the brackets are blue**, they might avoid making an incorrect assumption about what those brackets mean.

### Description and Requirements

Semantic highlighting is a feature defined by the Language Server Protocol, thus semantic tokens can be made available to all IDEs.

The token types that are used to label Mojo's square-bracket tokens might not be understood by the popular color themes. Thankfully, for VSCode in particular, it is possible for an extension to manually specify the default color that should be used for each semantic token type, either for specific themes, or irrespective of the theme. The color specification is done via the extension's `package.json`:
```
"contributes": {
    "configurationDefaults": {
        "editor.semanticTokenColorCustomizations": {
            "[<color theme>]": {
                 ...
            }
        }
    }
}
```

Whoever is working on the VSCode extension will understand what this means.

**Comment by DayDun** [comment]
Great suggestion. Though it does unfortunately clash with how VSCode by default colors matching braces:

<img width="358" alt="image" src="https://github.com/modularml/mojo/assets/14539902/9991abe1-9f0f-4a8a-86b9-595ae9303920">

Should maybe be toggleable in the extension so you can choose to either get colors based on semantics or based on matching braces.

Or maybe there's a clever compromise where you get the best of both worlds 🤷‍♂️

**Comment by DayDun** [comment]
Just realized the screenshotted code won't compile 🤦‍♂️

**Comment by nmsmith** [comment]
> Though it does unfortunately clash

That's true: the specific means of introducing the colorization in VSCode would require careful consideration. Though, not everyone uses the (non-semantic) bracket colorization feature seen in your screenshot. (Python programs typically don't involve many nested brackets.) Regardless, as you say, the semantic colorization only needs to be a configurable _option_. It doesn't need to be imposed upon anyone.

I should emphasise I am proposing two separate things here:
- For Mojo's language server to emit semantic tokens for each use of `[]` and `()`.
- For Mojo's VSCode extension to provide bracket colorization using these semantic tokens.

The first part can be implemented independently of concerns about colorization.

**Comment by rcghpge** [comment]
hey is this GitHub issue still going? I wanted to search the gh issues to see if anything popped up and this came up. I wanted to add dictionary literals into this discussion as well. 

`error: TODO: cannot emit dictionary literals yet
var keypad_dict = {}`
                  ^~
